certora/specs/Marketplace.spec

```diff
@@ -138,6 +153,52 @@ invariant slotMissedShouldBeEqualToNumberOfMissedPeriods(env e, Marketplace.Slot
 
+// STATUS - verified
+// cancelled request is always expired
+// https://prover.certora.com/output/6199/36b12b897f3941faa00fb4ab6871be8e?anonymousKey=de98a02041b841fb2fa67af4222f29fac258249f
+invariant cancelledRequestAlwaysExpired(env e, Marketplace.RequestId requestId)
+    currentContract.requestState(e, requestId) == Marketplace.RequestState.Cancelled =>
+        currentContract.requestExpiry(e, requestId) < lastBlockTimestampGhost;
+
+// STATUS - verified
+// failed request is always ended
+// https://prover.certora.com/output/6199/902ffe4a83a9438e9860655446b46a74?anonymousKey=47b344024bbfe84a649bd1de44d7d243ce8dbc21
+invariant failedRequestAlwaysEnded(env e, Marketplace.RequestId requestId)
+    currentContract.requestState(e, requestId) == Marketplace.RequestState.Failed =>
+        currentContract.requestContext(e, requestId).endsAt < lastBlockTimestampGhost;
+
+// STATUS - verified
+// finished slot always has finished request
+// https://prover.certora.com/output/6199/3371ee4f80354ac9b05b1c84c53b6154?anonymousKey=eab83785acb61ccd31ed0c9d5a2e9e2b24099156
+invariant finishedSlotAlwaysHasFinishedRequest(env e, Marketplace.SlotId slotId)
+    currentContract.slotState(e, slotId) == Marketplace.SlotState.Finished =>
+        currentContract.requestState(e, currentContract.slots(e, slotId).requestId) == Marketplace.RequestState.Finished;
+
+// STATUS - verified
+// paid slot always has finished or cancelled request
+// https://prover.certora.com/output/6199/6217a927ff2c43bea1124f6ae54a78fb?anonymousKey=d5e09d0d12658fd6b5f298fc04cea88da892c62d
+invariant paidSlotAlwaysHasFinishedOrCancelledRequest(env e, Marketplace.SlotId slotId)
+     currentContract.slotState(e, slotId) == Marketplace.SlotState.Paid =>
+         currentContract.requestState(e, currentContract.slots(e, slotId).requestId) == Marketplace.RequestState.Finished || currentContract.requestState(e, currentContract.slots(e, slotId).requestId) == Marketplace.RequestState.Cancelled
+    { preserved {
+        // ensures we start with a paid slot
+        require currentContract.slotState(e, slotId) == Marketplace.SlotState.Paid;
+      }
+    }
+
+// STATUS - verified
+// cancelled slot always belongs to cancelled request
+// https://prover.certora.com/output/6199/80d5dc73d406436db166071e277283f1?anonymousKey=d5d175960dc40f72e22ba8e31c6904a488277e57
+invariant cancelledSlotAlwaysHasCancelledRequest(env e, Marketplace.SlotId slotId)
+    currentContract.slotState(e, slotId) == Marketplace.SlotState.Cancelled =>
+        currentContract.requestState(e, currentContract.slots(e, slotId).requestId) == Marketplace.RequestState.Cancelled;
+
+// STATUS - verified
+// requests always ask for slot amount greaater than 0
+// https://prover.certora.com/output/6199/066a1cf8f74a457eaa820263d30f66be?anonymousKey=2217c966df8a9c32fe6112cd21fc0a8df601149f
+invariant requestAskSlotsGreaterThanZero(env e, Marketplace.RequestId requestId)
```

Comment:
Interesting, we are actually not validating that `request.ask.slots` is > 0, maybe we should add that. 
Does invariant "test" this? Or just constrains the prover to not test values <= 0?

Reply:
The invariant proofs that, for any given request, if the `ask.slots` are `> 0` they stay `> 0`.

It doesn't actually proof the case you've described. I'll provide a fix and try to write a rule that proofs it as well.
Good catch!

Reply:
> The invariant proofs that, for any given request, if the ask.slots are > 0 they stay > 0.

Also just realized that this invariant isn't very meaningful. Gonna remove it and add a test for `ask.slots > 0` check anyways.

Reply:
@AuHau added a separate PR for this fix here: https://github.com/codex-storage/codex-contracts-eth/pull/163